cloud-storage/cloud_storage.c

```diff
@@ -0,0 +1,909 @@
+
+#define BUFFER_SIZE 8192
+
+enum _operation_queue_item_type_t
+{
+   SYNC_FILES,
+   UPLOAD_FILE,
+   DOWNLOAD_FILE
+};
+
+struct _operation_queue_item;
+struct _operation_queue_item
+{
+   enum _operation_queue_item_type_t operation;
+
+   union
+   {
+      struct
+      {
+         folder_type_t folder_type;
+      } sync;
+
+      struct
+      {
+         folder_type_t folder_type;
+         char *local_file;
+      } upload;
+
+      struct
+      {
+         folder_type_t folder_type;
+         cloud_storage_item_t *remote_file;
+      } download;
+   } operation_parameters;
+
+   struct _operation_queue_item *next;
+};
+
+static bool _shutdown = true;
+static cloud_storage_provider_t *_provider;
+
+static char _game_states_config_value[8192] = "";
+
+static bool _need_sync_runtime_logs = false;
+
+cloud_storage_item_t *_game_states = NULL;
+
+static struct _operation_queue_item *_operation_queue = NULL;
+static slock_t *_operation_queue_mutex;
+
+static scond_t *_operation_condition;
+static sthread_t *_operation_thread;
+
+static void _sync_files_execute(folder_type_t folder_type);
+
+static void _upload_file_execute(folder_type_t folder_type, char *file_name);
+
+void cloud_storage_item_free(cloud_storage_item_t *item)
+{
+   cloud_storage_item_t *next;
+
+   while (item)
+   {
+      next = item->next;
+
+      free(item->id);
+      free(item->name);
+
+      if (item->item_type == CLOUD_STORAGE_FILE)
+      {
+         if (item->type_data.file.hash_value)
+         {
+            free(item->type_data.file.hash_value);
+         }
+         if (item->type_data.file.download_url)
+         {
+            free(item->type_data.file.download_url);
+         }
+      }
+
+      if (item->item_type == CLOUD_STORAGE_FOLDER && item->type_data.folder.children)
+      {
+         cloud_storage_item_free(item->type_data.folder.children);
+      }
+
+      free(item);
+
+      item = next;
+   }
+}
+
+static void _free_operation_queue(struct _operation_queue_item *queue)
+{
+   struct _operation_queue_item *next_item;
+
+   while (queue)
+   {
+      switch (queue->operation)
+      {
+         case UPLOAD_FILE:
+            if (queue->operation_parameters.upload.local_file)
+            {
+               free(queue->operation_parameters.upload.local_file);
+            }
+            break;
+         case DOWNLOAD_FILE:
+            if (queue->operation_parameters.download.remote_file)
+            {
+               cloud_storage_item_free(queue->operation_parameters.download.remote_file);
+            }
+            break;
+         default:
+            break;
+      }
+
+      next_item = queue->next;
+      free(queue);
+      queue = next_item;
+   }
+}
+
+static void _operation_thread_loop(void *user_data)
+{
+   while (!_shutdown)
+   {
+      struct _operation_queue_item *task;
+
+      slock_lock(_operation_queue_mutex);
+
+      scond_wait_timeout(_operation_condition, _operation_queue_mutex, 500);
```

Comment:
Why do you wake up the condition periodically? You signal it from the exit code, so it will shutdown gracefully and immediately. Right?

Reply:
I ran into an issue with the condition variable.
1. Add sync task to operation queue.
2. Signal the condition variable
3. The operation thread starts up and runs this function
4. Fails to wakeup when waiting on the condition variable, since the signal was already sent.

Thinking about this now, I should just skip the condtion wait if there are items in the queue.

Reply:
This has been improved as I mentioned above. Now it will only wait on the condition variable if the operation queue is empty. Also changed to only pull off the first item from the queue.